util/topgen/validate.py

```diff
@@ -255,14 +339,166 @@ def check_target(top, objs, tgtobj):
     return error, idxs
 
 
-def check_padctrl(top, prefix):
-    error = check_keys(top["padctrl"], padctrl_required, padctrl_optional,
-                       padctrl_added, prefix + " PadControl")
+def check_pinout(top, prefix):
+    error = check_keys(top['pinout'], pinout_required, pinout_optional,
+                       pinout_added, prefix + ' Pinout')
+
+    known_names = {}
+    for pad in top['pinout']['pads']:
+        error += check_keys(pad, pad_required, pad_optional,
+                            pad_added, prefix + ' Pinout')
+
+        # check name uniqueness
+        if pad['name'] in known_names:
+            log.warning('Pad name {} is not unique'.format(pad['name']))
+            error += 1
+        known_names.update({pad['name']: 1})
+
+        pad_type = pad.setdefault('type', top['pinout']['type_default'])
+        if not is_valid_pad_type(pad_type):
+            log.warning('Unkown pad type {}'.format(pad_type))
+            error += 1
+
+        pad_pwr = pad.setdefault('pwr', top['pinout']['pwr_default'])
+        if pad_pwr not in top['pinout']['pwr_domains']:
+            log.warning('Unkown pad pwr domain {}'.format(pad_type))
+            error += 1
+
+        pad.setdefault('muxed', top['pinout']['muxed_default'])
+        pad['muxed'], err = check_bool(pad['muxed'], prefix + ' Pinout')
+        error += err
+
     return error
 
 
+def check_io_name(name, io_type):
+    '''MIO/DIO/CIO name checker'''
+    assert (io_type in ['custom', 'muxed', 'dedicated'])
+    # In this case we allow <module>, <module>.<signal> or <module>.<signal>[<idx>]
+    if io_type in ['custom', 'muxed']:
+        pattern = r'^[a-zA-Z0-9_]+(\.[a-zA-Z0-9_]+(\[[0-9]*\]){0,1}){0,1}'
+    # Otherwise, module wildcards are not allowed and the full name must be given,
+    # i.e., either <module>.<signal> or <module>.<signal>[<idx>]
+    else:
+        pattern = r'^[a-zA-Z0-9_]+\.[a-zA-Z0-9_]+(\[[0-9]*\]){0,1}'
+    matches = re.match(pattern, name)
+    if matches is None:
+        log.warning('IO name {} ({}) has wrong format'.format(name, io_type))
```

Comment:
Is this really only a warning, and not an error? (here and below)

Reply:
This should probably have been an error - but this function has now been removed anyways.